services/jenkins/jenkins-coverage.service.js

```diff
@@ -1,80 +1,166 @@
-          }
-
-          try {
-            const coverageObject = json.instructionCoverage
-            if (coverageObject === undefined) {
-              badgeData.text[1] = 'inaccessible'
-              sendBadge(format, badgeData)
-              return
-            }
-            const coverage = coverageObject.percentage
-            if (isNaN(coverage)) {
-              badgeData.text[1] = 'unknown'
-              sendBadge(format, badgeData)
-              return
-            }
-            badgeData.text[1] = coverage.toFixed(0) + '%'
-            badgeData.colorscheme = coveragePercentageColor(coverage)
-            sendBadge(format, badgeData)
-          } catch (e) {
-            badgeData.text[1] = 'invalid'
-            sendBadge(format, badgeData)
-          }
+const jacocoCoverageSchema = Joi.object({
+  instructionCoverage: Joi.object({
+    percentage: Joi.number()
+      .min(0)
+      .max(100)
+      .required(),
+  }).required(),
+}).required()
+
+const coberturaCoverageSchema = Joi.object({
+  results: Joi.object({
+    elements: Joi.array()
+      .items(
+        Joi.object({
+          name: 'Lines',
+          ratio: Joi.number()
+            .min(0)
+            .max(100)
+            .required(),
         })
-      })
+      )
+      .min(1)
+      .required(),
+  }).required(),
+}).required()
+
+class BaseJenkinsCoverage extends BaseJsonService {
+  async fetch({ url, options, schema }) {
+    return this._requestJson({
+      url,
+      options,
+      schema,
+      errorMessages: {
+        404: 'job not found',
+      },
+    })
+  }
+
+  static render({ coverage }) {
+    return {
+      message: coverage.toFixed(0) + '%',
+      color: coveragePercentageColor(coverage),
+    }
+  }
+
+  static get defaultBadgeData() {
+    return { label: 'coverage' }
+  }
+
+  static get category() {
+    return 'build'
+  }
+
+  static buildUrl(scheme, host, job, plugin) {
+    return job.indexOf('/') > -1
+      ? `${scheme}://${host}/${job}/lastBuild/${plugin}/api/json`
+      : `${scheme}://${host}/job/${job}/lastBuild/${plugin}/api/json`
+  }
+
+  static buildOptions(treeParam) {
+    const options = {
+      qs: {
+        tree: treeParam,
+      },
+    }
+    if (serverSecrets && serverSecrets.jenkins_user) {
+      options.auth = {
+        user: serverSecrets.jenkins_user,
+        pass: serverSecrets.jenkins_pass,
+      }
+    }
+    return options
+  }
+}
+
+class JacocoJenkinsCoverage extends BaseJenkinsCoverage {
+  async handle({ scheme, host, job }) {
+    const url = this.constructor.buildUrl(scheme, host, job, 'jacoco')
+    const options = this.constructor.buildOptions(
+      'instructionCoverage[percentage]'
     )
+    const json = await this.fetch({
+      url,
+      options,
+      schema: jacocoCoverageSchema,
+    })
+    return this.constructor.render({
+      coverage: json.instructionCoverage.percentage,
+    })
+  }
+
+  static get url() {
+    return {
+      base: 'jenkins/j',
```

Comment:
The old badges supported an optional `-ci` in the URL: `/^\/jenkins(?:-ci)?\/(c|j)\/(http(?:s)?)\/([^/]+)\/(.+)\.(svg|png|gif|jpg|json)$/`. 

This was designed from the start in 6f616fc58d2c7aa6ea25292f45268839157bf9c7, but was never advertised in the homepage examples. Unless I'm missing something it's most likely that no one is using it, unless they've dug into the code and gone "cool, I can make my badges URL longer with an optional useless bit". Therefore I think we should drop that bit for consistency with the other Jenkins badges and also because I'm unsure how we would translate this possibility in the new architecture (would it entail making the base empty and moving it to the format parameter?).

Reply:
I did a code search on GitHub.

* `"https://img.shields.io/jenkins"`: 837 results
* `"https://img.shields.io/jenkins-ci"`: 0 results

I reckon we can probably bin it :wastebasket: